yt_dlp/extractor/bilibili.py

```diff
@@ -1,509 +1,391 @@
     }]
 
-    _APP_KEY = 'iVGUTjsxvpLeuDCf'
-    _BILIBILI_KEY = 'aHRmhWMLkdeMuILqORnYZocwMBpMEOdt'
-
-    def _report_error(self, result):
-        if 'message' in result:
-            raise ExtractorError('%s said: %s' % (self.IE_NAME, result['message']), expected=True)
-        elif 'code' in result:
-            raise ExtractorError('%s returns error %d' % (self.IE_NAME, result['code']), expected=True)
-        else:
-            raise ExtractorError('Can\'t extract Bangumi episode ID')
-
     def _real_extract(self, url):
-        url, smuggled_data = unsmuggle_url(url, {})
-
         mobj = self._match_valid_url(url)
-        video_id = mobj.group('id_bv') or mobj.group('id')
-
-        av_id, bv_id = self._get_video_id_set(video_id, mobj.group('id_bv') is not None)
-        video_id = av_id
+        video_id = mobj.group('bv_id') or mobj.group('aid')
 
-        info = {}
-        anime_id = mobj.group('anime_id')
-        page_id = mobj.group('page')
         webpage = self._download_webpage(url, video_id)
-
-        # Bilibili anthologies are similar to playlists but all videos share the same video ID as the anthology itself.
-        # If the video has no page argument, check to see if it's an anthology
-        if page_id is None:
+        initial_state = self._search_json(r'window.__INITIAL_STATE__\s*=\s*', webpage, 'initial state', video_id)
+
+        video_data = traverse_obj(initial_state, 'videoData') or {}
+        bv_id = video_data['bvid']
+        cid = video_data.get('cid')
+
+        page_list_json = traverse_obj(
+            self._download_json(
+                'https://api.bilibili.com/x/player/pagelist', video_id,
+                fatal=False, query={'bvid': bv_id, 'jsonp': 'jsonp'},
+                note='Extracting videos in anthology'),
+            'data', expected_type=list) or []
+        has_multi_p = len(page_list_json or []) > 1
+        part_id = int_or_none(parse_qs(url).get('p', [None])[-1])
+
+        title = video_data.get('title')
+
+        if has_multi_p and part_id is None:
+            # Bilibili anthologies are similar to playlists but all videos share the same video ID as the anthology itself.
+            # If the video has no page argument and it's an anthology, download as a playlist
             if not self.get_param('noplaylist'):
-                r = self._extract_anthology_entries(bv_id, video_id, webpage)
-                if r is not None:
+                ret = self.playlist_from_matches(page_list_json, bv_id, title, ie=BiliBiliIE,
+                                                 getter=lambda entry: f'https://www.bilibili.com/video/{bv_id}?p={entry["page"]}')
+                if ret is not None:
                     self.to_screen('Downloading anthology %s - add --no-playlist to just download video' % video_id)
-                    return r
+                    return ret
             else:
                 self.to_screen('Downloading just video %s because of --no-playlist' % video_id)
```

Comment:
Note: take care that the video_id of anthology entries don't overlap. Also, ensure that the id remains same whether --no-playlist is being used or not

Reply:
ok, video_id of anthology entries is like `xxxx_partXX`, anthology id is `xxxx`